src/main/java/com/streamr/client/Subscription.java

```diff
@@ -49,4 +73,96 @@ public void setState(State state) {
     public MessageHandler getHandler() {
         return handler;
     }
+
+    public boolean isResending() {
+        return resending;
+    }
+
+    public void setResending(boolean resending) {
+        this.resending = resending;
+    }
+
+    public ResendRangeRequest handleMessage(StreamMessage msg, String sessionToken) {
+        return handleMessage(msg, sessionToken, false);
+    }
+
+    public ResendRangeRequest handleMessage(StreamMessage msg, String sessionToken, boolean isResend) {
+        String key = msg.getPublisherId() + msg.getMsgChainId();
+        if (resending && !isResend) {
+            queue.add(msg);
+            return null;
+        } else if (checkForGap(msg.getPreviousMessageRef(), key) && !resending) {
+            queue.add(msg);
+            MessageRef from = lastReceivedMsgRef.get(key); // cannot know the first missing message so there will be a duplicate received
+            MessageRef to = msg.getPreviousMessageRef();
+            return new ResendRangeRequest(msg.getStreamId(), msg.getStreamPartition(), id, from, to, msg.getPublisherId(), msg.getMsgChainId(), sessionToken);
+        } else {
+            MessageRef msgRef = msg.getMessageRef();
+            Integer res = null;
+            MessageRef last = lastReceivedMsgRef.get(key);
+            if (last != null) {
+                res = msgRef.compareTo(last);
+            }
+            if (res != null && res <= 0) {
+                log.debug(String.format("Sub %s already received message: %s, lastReceivedMsgRef: %s. Ignoring message.", id, msgRef, last));
+            } else {
+                lastReceivedMsgRef.put(key, msgRef);
+                getHandler().onMessage(this, msg);
+            }
+            return null;
+        }
+    }
+
+    public ResendRangeRequest handleQueue(String sessionToken) {
+        ResendRangeRequest gap = null;
+        while(!queue.isEmpty() && gap == null) {
+            StreamMessage msg = queue.poll();
+            gap = handleMessage(msg, sessionToken);
+        }
+        return gap;
+    }
+
+    public void handleError(Exception e, StreamMessage msg) {
+        String key = msg.getPublisherId() + msg.getMsgChainId();
+        if (e instanceof IOException) {
```

Comment:
(Relates to a previous comment)
- We could check that the message is ok before calling the message handler, and deal with the situation within this class instead of catching it in `StreamrClient` and calling `handleError`
- In `handleMessage`, we're actually updating `lastReceivedMsgRef` before calling the message handler, so the `lastReceivedMsgRef` should be updated even if `IOException` occurs within the message handler call?